Source/WebInspectorUI/UserInterface/Views/FontVariationDetailsSectionRow.js

```diff
@@ -142,6 +165,87 @@ WI.FontVariationDetailsSectionRow = class FontVariationDetailsSectionRow extends
 
         return 1;
     }
+
+    _showValidity()
+    {
+        if (!this._hasValidValue)
+            this.warningMessage = WI.UIString("Axis value outside of supported range: %s – %s", "A warning that is shown in the Font Details Sidebar when the value for a variation axis is outside of the supported range of values").format(this._formatAxisValueAsString(this._minimumValue), this._formatAxisValueAsString(this._maximumValue));
+        else
+            this.warningMessage = null;
+    }
+
+    _handleValueBlurEvent(event)
+    {
+        this._skipUpdatingInputValue = false;
+        this.value = this._value;
+    }
+
+    _handleValueInputEvent(event)
+    {
+        if (this._showValidityTimeout)
+            clearTimeout(this._showValidityTimeout);
+
+        // Allow the input to temporarily hold an invalid value without immediately warning the user.
+        this._showValidityTimeout = setTimeout(this._showValidity.bind(this), 750);
+
+        let valueAsString = event.target.value;
+        let valueAsNumber = event.target.valueAsNumber;
+
+        this._hasValidValue = false;
+        const validValueFormat = /\-?\d+(.\d+)?/g;
+
+        if (!validValueFormat.test(valueAsString)) {
+            this._skipUpdatingInputValue = true;
+
+            this.dispatchEventToListeners(WI.FontVariationDetailsSectionRow.Event.VariationValueChanged, {tag: this._tag, value: this._defaultValue});
+            return;
+        }
+
+        if (!(this._minimumValue <= valueAsNumber && valueAsNumber <= this._maximumValue)) {
+            this._skipUpdatingInputValue = true;
+
+            this.dispatchEventToListeners(WI.FontVariationDetailsSectionRow.Event.VariationValueChanged, {tag: this._tag, value: Math.min(Math.max(this._minimumValue, valueAsNumber), this._maximumValue)});
+            return;
+        }
+
+        this._hasValidValue = true;
+        this._skipUpdatingInputValue = false;
+        this.warningMessage = null;
+        this.value = valueAsNumber;
+
+        clearTimeout(this._showValidityTimeout);
+        this.dispatchEventToListeners(WI.FontVariationDetailsSectionRow.Event.VariationValueChanged, {tag: this._tag, value: valueAsNumber});
+    }
+
+    _handleValueKeydownEvent(event)
+    {
+        // By default, a HTML number input will accept and display any character.
+        if (event.key.length === 1 && !/[0-9\.\-]/.test(event.key)) {
+            event.preventDefault();
+            return;
+        }
+
+        let valueAsNumber = event.target.valueAsNumber;
+        let step = this._step;
+
+        if (event.shiftKey && event.altKey)
```

Comment:
i coulda sworn we had a utility for this logic somewhere, but i cant really find it :(

we definitely should tho.  maybe a followup?

Reply:
There are two other places where the modifier keys change the step:

- [BoxShadowEditor.js#L303-L324](https://github.com/WebKit/WebKit/blob/main/Source/WebInspectorUI/UserInterface/Views/BoxShadowEditor.js#L303-L324)
Operating on an `<input type="text">`
- [SpreadsheetTextField.js#L285-L295](https://github.com/WebKit/WebKit/blob/main/Source/WebInspectorUI/UserInterface/Views/SpreadsheetTextField.js#L285-L295)
Operating on a `<span contenteditable>`

Neither is working on a `<input "type=number">` that natively handles increment/decrement, and which we're trying to use here.

They're not consistent either, one assumes a default step of `1`. For the variation axis, the default step can be `0.1` or `0.01` for axes with very short ranges ([the Skia font](https://v-fonts.com/fonts/skia) is a good example).

As a follow-up we're considering making a custom element that wraps the native HTML number input with the behaviors and constraints we expect in Web Inspector. It might not replace the `<span contenteditable>` case which handles a lot of extra behavior, but it can serve as a predictable number input for other uses, like this one in the Fonts sidebar panel.

Reply:
i was more just referring to the "look at `event.shiftKey` and multiply by `10`" parts, but i love the idea of making a generalized thing 😃

Reply:
also btw there's also this logic inside `WI.startEditing`, which i think is used for DOM editing?  im pretty sure i added support for this kinda stuff in the DOM tree too